### PR TITLE
Coach can create and start videocall

### PR DESCRIPTION
**Pull request by HannaSwDn**
# PT Link

[Pivotal Tracker](https://www.pivotaltracker.com/story/show/163688955)

# User Story

```
As a coach
In order to have a session
I would like to have a video call automatically added when a session is created
```

# Screenshots

gif

# Description of PR

- Implements Agora videocall
- Adds start button to booked sessions for the coach to be able to start a video call

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -3,6 +3,99 @@
 //= require turbolinks
 //= require_tree .
 
+// videocall
+
+let remoteContainer;
+let canvasContainer;
+
+document.addEventListener('turbolinks:load', event => {
```
Comment (kianaditya) [inline app/assets/javascripts/application.js]: We are using the same function for Stripe down below (line 128), please use one function for both !

File: app/controllers/videocall_controller.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class VideocallController < ApplicationController
```
Comment (kianaditya) [inline app/controllers/videocall_controller.rb]: Why do we need this ?

File: app/assets/javascripts/application.js (hunk 1 above)
Comment (HannaSwDn) [reply app/assets/javascripts/application.js]: Good point, fixed!

File: app/controllers/videocall_controller.rb (hunk 2 above)
Comment (HannaSwDn) [reply app/controllers/videocall_controller.rb]: We do not need the index action, removed. However we do need the controller for routing.

Comment (HannaSwDn) [reply app/controllers/videocall_controller.rb]: We tried to remove it but got uninitialized controller error.

File: app/controllers/videocall_controller.rb [hunk 3]
```diff
@@ -0,0 +1,3 @@
...
```
Comment (GergKllai1) [inline app/controllers/videocall_controller.rb]: Do you still need this?

File: app/views/videocall/index.html.erb [hunk 4]
```diff
@@ -0,0 +1 @@
+<%= react_component("StartSession", props: @session_props, prerender: Rails.env.test?) %>
```
Comment (GergKllai1) [inline app/views/videocall/index.html.erb]: If you have set up routing in app.jsx you don't need that. You are rendering it in main.html.erb not here.
Try removing it

File: config/routes.rb [hunk 5]
```diff
@@ -6,4 +6,6 @@
 	match 'register', to: 'main#index', via: [:get]
   match 'login', to: 'main#index', via: [:get]
   match 'addsession', to: 'main#index', via: [:get, :post]
+  resources :videocall, only: [:index]
```
Comment (GergKllai1) [inline config/routes.rb]: Remove this we are not using it anymore

File: features/step_definitions/assertion_steps.rb [hunk 6]
```diff
@@ -12,6 +12,10 @@
     expect(page).to have_no_content content
 end
 
+Then("I should be on the videocall page") do
```
Comment (GergKllai1) [inline features/step_definitions/assertion_steps.rb]: You are not routing to this page anymore

File: features/coach_can_create_and_start_videocall.feature [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        And the following sessions exist
+            | title    | start_date          | end_date            | status |
+            | Crossfit | 2019-02-01 15:00:00 | 2019-02-01 15:30:00 | booked |
+
+        And I am logged in as 'coach@email.com'
+        And I visit the site
+
+    Scenario: Videocall gets added to session
+        Given I click 'Start'
+        Then I should be on the videocall page
```
Comment (GergKllai1) [inline features/coach_can_create_and_start_videocall.feature]: You are not routing to this page anymore

File: features/trainee_can_buy_a_session.feature [hunk 8]
```diff
@@ -7,16 +7,17 @@ Feature: User can buy a subscription
...
 		And the following sessions exist
-			| title     | start_date          | end_date            | status    |
-			| Crossfit  | 2019-02-01 15:00:00 | 2019-02-01 15:30:00 | available |
+			| title    | start_date          | end_date            | status    |
+			| Crossfit | 2019-02-01 15:00:00 | 2019-02-01 15:30:00 | available |
 		And the time is 2019 1 30 10:00:00
 
 	Scenario: Trainee can successfully buy a subscription
 		And I am logged in as 'john@mail.com'
+		Then show me the page
```
Comment (GergKllai1) [inline features/trainee_can_buy_a_session.feature]: Why do we have this here?
Remove it if it has no purpose

File: app/assets/javascripts/application.js [hunk 9]
```diff
@@ -3,6 +3,94 @@
...
+	let ctx = canvas.getContext('2d');
+	let video = document.getElementById(`video${streamId}`);
+
+	video.addEventListener('loadedmetadata', function () {
+		canvas.width = video.videoWidth;
+		canvas.height = video.videoHeight;
+	});
+
+	video.addEventListener('play', function () {
+		var $this = this;
```
Comment (GergKllai1) [inline app/assets/javascripts/application.js]: This seems old syntax, can you use `let`?

File: app/assets/javascripts/application.js [hunk 10]
```diff
@@ -3,6 +3,94 @@
...
+	codec: "h264"
+});
+
+client.init("5b4594eea0de44a39e6a3ce3605d27b8", () => console.log("AgoraRTC client initialized"), handleFail);
+
+client.join(null, "any-channel", null, (uid) => {
+
+	let localStream = AgoraRTC.createStream({
+		streamID: uid,
+		audio: false,
```
Comment (GergKllai1) [inline app/assets/javascripts/application.js]: What does this do? Doesn't that mean that we have no audio?

File: app/assets/javascripts/application.js [hunk 11]
```diff
@@ -33,6 +121,8 @@ const initiateStripe = () => {
 };
 
 document.addEventListener('turbolinks:load', () => {
+	remoteContainer = document.getElementById("remote-container");
+    canvasContainer = document.getElementById("canvas-container");
```
Comment (GergKllai1) [inline app/assets/javascripts/application.js:125]: Indentaiton

**Review by GergKllai1 — CHANGES_REQUESTED** [review]
All in all mighty impressed guys :)

File: app/controllers/videocall_controller.rb (hunk 3 above)
Comment (johannaanden) [reply app/controllers/videocall_controller.rb]: No, removed

File: app/views/videocall/index.html.erb (hunk 4 above)
Comment (johannaanden) [reply app/views/videocall/index.html.erb]: Removed

File: config/routes.rb (hunk 5 above)
Comment (HannaSwDn) [reply config/routes.rb]: Done

File: features/step_definitions/assertion_steps.rb (hunk 6 above)
Comment (HannaSwDn) [reply features/step_definitions/assertion_steps.rb]: Done

File: features/coach_can_create_and_start_videocall.feature (hunk 7 above)
Comment (HannaSwDn) [reply features/coach_can_create_and_start_videocall.feature]: Done

File: features/trainee_can_buy_a_session.feature (hunk 8 above)
Comment (HannaSwDn) [reply features/trainee_can_buy_a_session.feature]: Done

File: app/assets/javascripts/application.js (hunk 9 above)
Comment (HannaSwDn) [reply app/assets/javascripts/application.js]: Done

File: app/assets/javascripts/application.js (hunk 10 above)
Comment (HannaSwDn) [reply app/assets/javascripts/application.js]: We are not 100% sure before we can see it on Heroku, but it means we have no audio track for the videocall. We have changed it true for now.

**Review by kianaditya — APPROVED** [review]
LGTM!

**Review by GergKllai1 — APPROVED** [review]

**Pull request merged**